edgelet/iotedged/src/lib.rs

```diff
@@ -856,13 +856,17 @@ fn manual_provision(
     tokio_runtime.block_on(provision)
 }
 
-fn external_provision_payload(key: &str) -> (DerivedKeyStore<MemoryKey>, MemoryKey) {
-    let memory_key = MemoryKey::new(key);
+fn external_provision_payload(key: &str) -> Result<(DerivedKeyStore<MemoryKey>, MemoryKey), Error> {
+    let memory_key = MemoryKey::new(base64::decode(&key).map_err(|_| {
+        Error::from(ErrorKind::Initialize(
+            InitializeErrorReason::ExternalProvisioningClient,
+        ))
+    })?);
```

Comment:
Should the above be the following instead?
```rust
let memory_key = MemoryKey::new(base64::decode(&key).map_err(|err| {
    Error::from(err.context(ErrorKind::Initialize(
        InitializeErrorReason::ExternalProvisioningClient,
    )))
})?);
```

Reply:
I thought so too but looking at a similar call to decode the base64 SAS key for 'manual' mode in the method parse_device_connection_string, I noticed this pattern where we're ignoring the error originating from the decode call. 

`let key = MemoryKey::new(base64::decode(&key).map_err(|_| {
            ParseManualDeviceConnectionStringError::MalformedParameter(SHAREDACCESSKEY_KEY)
        })?);`

Reply:
I see. Yeah, I don't feel too strongly about this. I am OK with leaving it as-is.

Just that with the context, the user is likely to know that the problem is with the key part of the provisioning result. It is arguably debatable as to whether divulging that information is a good idea from a security perspective. The `parse_device_connection_string` function does seem to let you know what part of the connection string was found unacceptable.

Reply:
Yes, it intentionally suppresses the original error to avoid leaking the key in system logs.

Reply:
I see. I could log a message where this error is being returned similar to other places where we handle invalid data sent in the response of the call to get provisioning information such as this:

`info!("Unexpected auth type. Only symmetric keys are expected");
                        return Err(Error::from(ErrorKind::Initialize(
                            InitializeErrorReason::ExternalProvisioningClient,
                        )));`

What do you think?

Reply:
@shantanu1singh You might want to have a nested `Reason` in your `InitializeErrorReason::ExternalProvisioningClient` for what exactly failed when creating the external provisioning client. I see `InitializeErrorReason::ExternalProvisioningClient` used in seven different locations already, so you / the user will probably not find it easy to determine which of these eight caused the failure.

This will also solve Raj's concern about at least revealing that the malformed key was the reason for the failure.

Reply:
Got it, I'll get this change in cause I'm trying to fix this issue asap. I'll send another PR sometime this week to add nested Reasons.